### ISSUE TITLE
SCAFFOLD

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2014]
# SCAFFOLD

* **Title:** SCAFFOLD: Stochastic Controlled Averaging for Federated Learning
* **Venue:** ICML 2020
* **Link to paper:** https://proceedings.mlr.press/v119/karimireddy20a.html

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [ ] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by jafermarq** [comment]
Hi Aashish (@aashishkolluri ), this looks like a good plan for `SCAFFOLD`! Am I understanding it correctly that when you say "rows 2 and 3" for Table 3 you mean the block of rows for `FedAvg` and for `SCAFFOLD` in the table? Also, could you confirm were you are based is on the list of countries/territories that are _eligible_? You can find the list afterthe second point in Step 2 (you don't have to say where you are, just confirming that you are eligible is good enough)

**Comment by aashishkolluri** [comment]
Yes you are right about FedAvg and SCAFFOLD. I am based on the list of countries.

**Comment by jafermarq** [comment]
Perfect! I have ✅ all the points in Step 1 & 2 since they are now completed. I have also added you as the _assignee_ to this issue and labeled it as _In Progress_. You can find info on how to start with the code by following the link in the _what happens next?_ section after Step 2 in the description of this issue. If you have some doubts or questions please preferably use our `#questions` or `#summer-of-reproducibility` Slack channel.

Looking forward to seeing `SCAFFOLD` in action!

**Comment by jafermarq** [comment]
Hi @aashishkolluri !
This is just a gentle reminder that the Flower *Summer of Reproducibility* is ending at the end of the month. With just a little more than 3 weeks to go, we are excited to see quite a few baselines well ahead in the process with [their respective PRs close to ready.](https://github.com/adap/flower/pulls?q=is%3Apr+is%3Aopen+label%3Asummer-of-reproducibility) If your PR is already on the list, great !! Please make sure the PR is linked to this issue (you just need to copy the URL of this issue somewhere in the main message of your PR). Ping me when you'd like me to take a look.

Also, make sure you keep an eye:eyes: on the **#summer-of-reproducibility** channel in the Flower Slack. I’ll announce very soon a new (the third!) round of 1:1 ask-me-anything sessions to help _Summer of Reproducibility_ contributors like yourself to meet the deadline. Please consider booking a time slot if you want to chat with me about your baseline, potential issues you have making your code run, how to open a PR, doubts about what to include in your readme, how to use Hydra configs more effective, etc … all questions are welcome!!

**Comment by BalajiAI** [comment]
Hi @jafermarq, can I take this one up?
I've aldready implemented SCAFFOLD - see [here](https://github.com/BalajiAI/Federated-Learning/tree/main). So I just need to make few changes to make it compatible with Flower.

**Comment by wittenator** [comment]
I am currently implementing SCAFFOLD as a baseline here: https://github.com/wittenator/flower/tree/rework_fedprox_baseline/baselines/scaffold

**Comment by BalajiAI** [comment]
> Hi @wittenator, @BalajiAI,
> 
> Just checking in here. Are you still eager to complete this baseline implementation?
> 
> Best regards William

Yes, @WilliamLindskog. I've some free time in this weekend, so I can able to complete it in a week or two.

**Comment by wittenator** [comment]
I implemented the method in https://github.com/wittenator/flower/tree/rework_fedprox_baseline , but I currently don't have time to do the reproduction since the paper has its own way of splitting data and own non-standard models.

**Comment by WilliamLindskog** [comment]
Hi @BalajiAI 

Just checking in how the baseline is going and if there is anything you wonder about.

**Comment by WilliamLindskog** [comment]
Hi @BalajiAI, 

Just checking in. Do you have any updates to this baseline?

**Comment by wittenator** [comment]
I currently don't have time to implement the peculiarities of the SCAFFOLD paper sadly. I did use the code for a project though and the results are in line what other people report in their result sections, so maybe that it is a good starting point for somebody who wants to do the reproduction itself.

**Comment by WilliamLindskog** [comment]
Migrating this to [Flower Discuss](https://discuss.flower.ai/t/federated-baseline-condtributions/602) for now.
